File: ros2_ws/src/actions_cpp/src/count_until_server.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include "my_robot_interfaces/action/count_until.hpp"

using CountUntil = my_robot_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ServerGoalHandle<CountUntil>;
using namespace std::placeholders;


class CountUntilServerNode : public rclcpp::Node
{
public:
    CountUntilServerNode() : Node("count_until_server")
    {
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        count_until_server_ = rclcpp_action::create_server<CountUntil>(
            this,
            "count_until",
            std::bind(&CountUntilServerNode::goal_callback, this, _1, _2),
            std::bind(&CountUntilServerNode::cancel_callback, this, _1),
            std::bind(&CountUntilServerNode::handle_accepted_callback, this, _1),
            rcl_action_server_get_default_options(),
            cb_group_
        );
        RCLCPP_INFO(this->get_logger(), "Action server has been started.");
    }

private:
    // Class attributes:
    rclcpp_action::Server<CountUntil>::SharedPtr count_until_server_;
    rclcpp::CallbackGroup::SharedPtr cb_group_;
    std::mutex mutex_; // For lock variable when threading
    std::shared_ptr<CountUntilGoalHandle> goal_handle_;
    rclcpp_action::GoalUUID preemeted_goal_id_;

    // Functions
    rclcpp_action::GoalResponse goal_callback(const rclcpp_action::GoalUUID & uuid, std::shared_ptr<const CountUntil::Goal> goal)
    {
        (void)uuid; // To avoid build warnings for unused parameters
        // (void)goal; // To avoid build warnings for unused parameters

        RCLCPP_INFO(this->get_logger(), "Received a goal.");

        // // Policy: Refuse new goal if a goal is already active.
        // {
        //     std::lock_guard<std::mutex> lock(mutex_);
        //     // If another thread is trying to access the variables in the spoce ({}),
        //     // its going to need to wait for this one to finish.
        //     if (goal_handle_){
        //         if (goal_handle_->is_active()){
        //             RCLCPP_INFO(this->get_logger(), "A goal is still active, Rejecting new goal.");
        //             return rclcpp_action::GoalResponse::REJECT;
        //         }
        //     }
        // }


        if (goal->target_number <= 0.0) {

            RCLCPP_INFO(this->get_logger(), "This is not a positive number. Rejecting the goal.");
            return rclcpp_action::GoalResponse::REJECT;
        }

        // Policy: preemnt existing goal when receiveing new valid goal
        {
            std::lock_guard<std::mutex> lock(mutex_);
            if (goal_handle_){
                if (goal_handle_->is_active()){
                    RCLCPP_INFO(this->get_logger(), "Aborting current goal and accepting new goal.");
                    preemeted_goal_id_ = goal_handle_->get_goal_id();
                }
                // Now we can use the "preemeted_goal_id_" in execute_goal func.
            }
        }

        RCLCPP_INFO(this->get_logger(), "Accepting the goal.");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse cancel_callback(const std::shared_ptr<const CountUntilGoalHandle> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Received a cancel request.");
        (void)goal_handle; // To avoid build warnings for unused parameters
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted_callback(const std::shared_ptr<CountUntilGoalHandle> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Executing the goal.");
        execute_goal(goal_handle);
    }

    void execute_goal(const std::shared_ptr<CountUntilGoalHandle> goal_handle)
    {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            // If another thread is trying to access the variables in the spoce ({}),
            // its going to need to wait for this one to finish.
            this->goal_handle_ = goal_handle;
            // In order to Refuse new goal when one executed we keep the goal handle in this attribute.
        }
        
        // get request from goal
        int target_number = goal_handle->get_goal()->target_number;
        double period = goal_handle->get_goal()->period;

        // Execute the action
        int counter =0;
        auto result = std::make_shared<CountUntil::Result>();
        auto feedback  = std::make_shared<CountUntil::Feedback>();
        rclcpp::Rate loop_rate(1.0/period);
        for (int i = 0; i < target_number; i++){
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if (goal_handle->get_goal_id() == preemeted_goal_id_){
                    result->reached_number = counter;
                    goal_handle->abort(result);
                    return;
                }
            }
            
            if (goal_handle->is_canceling()){ // Check if cancelled

                result->reached_number = counter;
                goal_handle->canceled(result);
                return;
            }

            counter++;
            RCLCPP_INFO(this->get_logger(), "%d", counter);

            feedback->current_number = counter;
            goal_handle->publish_feedback(feedback);
            loop_rate.sleep();
        }

        // Set the final state and return result
        result->reached_number = counter;
        goal_handle->succeed(result);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    // rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/actions_py/actions_py/count_until_server.py
#!/usr/bin/env python3
import rclpy
import time
import threading # To use goal lock
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle
from my_robot_interfaces.action import CountUntil
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup


class CountUntilServerNode(Node):
    def __init__(self):
        super().__init__("count_until_server")
        
        self.goal_handle_: ServerGoalHandle = None    # initiating an empty goal handle
        self.goal_lock_ = threading.Lock()   # Search about threading library and thread locking.
        self.goal_queue_ = [] # Goal queue list
        
        self.count_until_server_ = ActionServer(
            self,
            CountUntil,
            "count_until",
            goal_callback = self.goal_callback,
            handle_accepted_callback = self.handle_accepted_callback,
            cancel_callback = self.cancel_callback,
            execute_callback = self.execute_callback,
            callback_group = ReentrantCallbackGroup()
        )
        self.get_logger().info("Action Server has been started.")
        
    def goal_callback(self, goal_request: ServerGoalHandle):
        
        self.get_logger().info("Received a goal.")
        
        # # Policy: refuse new goal if current goal is still active
        # with self.goal_lock_: # making sure this variable is being accessed once at a time.
        #     if self.goal_handle_ is not None and self.goal_handle_.is_active:
        #         self.get_logger().info("A goal is already active, Rejecting new goal.")
        #         return GoalResponse.REJECT
        
        
        # Validate the goal request
        if goal_request.target_number <= 0:
            self.get_logger().info("Rejecting goal. Invalid target number.")
            return GoalResponse.REJECT
        
        # # Policy: preempt existing goal when receiving new goal
        # with self.goal_lock_:
        #     if self.goal_handle_ is not None and self.goal_handle_.is_active:
        #         self.get_logger().info("Abort current goal and access new goal.")
        #         self.goal_handle_.abort()
                
            
        self.get_logger().info("Accepting goal.")
        return GoalResponse.ACCEPT
    
    def handle_accepted_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            if self.goal_handle_ is not None:
                self.goal_queue_.append(goal_handle)
                self.get_logger().info("Another goal is being executed at the moment.\nThe goal has been added the queue and will be executed when the current executing is finished.")
            else:
                goal_handle.execute()
    
    def cancel_callback(self, goal_handle: ServerGoalHandle):
        
        self.get_logger().info("Received a cancel request.")
        return CancelResponse.ACCEPT # or REJECT
        
    def execute_callback(self, goal_handle: ServerGoalHandle):
        
        with self.goal_lock_: # making sure this variable is being accessed once at a time.
            self.goal_handle_ = goal_handle
        
        # Get request from goal
        target_number = goal_handle.request.target_number
        period = goal_handle.request.period
        
        # Execute the action
        self.get_logger().info("Executing the goal...")
        feedback = CountUntil.Feedback()
        result = CountUntil.Result()
        counter = 0
        for i in range(target_number):
            if not goal_handle.is_active: # keep checking if goal was interrupted and aborted.
                result.reached_number = counter
                self.process_next_goal_in_queue() # check for queue goals.
                return result
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Canceling the goal.")
                goal_handle.canceled()
                result.reached_number = counter
                self.process_next_goal_in_queue() # check for queue goals.
                return result
            counter +=1
            self.get_logger().info(str(counter))
            feedback.current_number = counter
            goal_handle.publish_feedback(feedback)
            time.sleep(period)
            
        # Once done, set goal final state
        goal_handle.succeed()
        
        # If aborted
        # goal_handle.abort()
        
        # and send the result
        result = CountUntil.Result()
        result.reached_number = counter
        self.process_next_goal_in_queue() # check for queue goals.
        return result
    
    def process_next_goal_in_queue(self):
        with self.goal_lock_:
            if len(self.goal_queue_) > 0:
                self.goal_queue_.pop(0).execute() # Remove the element for the list and execute.
            else:
                self.goal_handle_ = None
            

def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServerNode()
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()


if __name__ == "__main__":
    main()
